admin/app/components/solidus_admin/sidebar/item/component.rb

```diff
@@ -74,4 +76,12 @@ def nested_items
   def active?
     @item.active?(@url_helpers, request.fullpath)
   end
+
+  private
+
+  # We have the request object injected on initialization for testing purposes.
+  # However, the actual Rails request object is only available at render time.
+  def before_render
+    @request ||= request
+  end
```

Comment:
This smells like an indication that the component can be dumbed down a little bit. 

A good technique that yielded good results in the past for me was to only accept values and use a class method to build the component starting from a more complex object.

In this case it could look like this:

```rb
  def self.from_item(item, request, url_helpers)
    new(
      key: item.key,
      icon: item.icon,
      href: item.path(url_helpers)
      active: item.active?(url_helpers, request.fullpath)
      top_level: item.top_level,
      children: item.children.map { |child| from_item(child, request, url_helpers) }
    )
  end
```

Which would remove the need to provide `request`, `url_helpers` and all that jazz to the initializer and have a preview which is really easy to play with and demonstrate in different states.

Reply:
Thanks for raising that, @elia. I agree, there's something we can extract about how the item deals with the request and URL generation contexts. I already had it at the back of my mind, but I'd like to think more about it and, in any case, tackle it in the context of a separate PR. I want to move some of the stuff to the main navigation configuration.

For now, I revisited the stuff and removed the `before_render` "hack" by just passing around the whole path string instead of the entire request object.

Does it work for you?

Reply:
> I already had it at the back of my mind, but I'd like to think more about it and, in any case, tackle it in the context of a separate PR. I want to move some of the stuff to the main navigation configuration.

That's awesome, you have the best knowledge about where to move the logic, and specs and previews always benefit from accepting simpler objects. Tackling it in a separate PR works great! 

> For now, I revisited the stuff and removed the `before_render` "hack" by just passing around the whole path string instead of the entire request object.

Great call 🙌